SysVr2.0_32000/SysVr2.0_32000/man/u_man/man1/crontab.1
'\" t
.TH CRONTAB 1
.SH NAME
crontab - user crontab file
.SH SYNOPSIS
.B crontab
[file] 
.br
.B crontab
\fB-r\fR
.br
.B crontab
\fB-l\fR
.SH DESCRIPTION
.I Crontab
copies the specified file,
or standard input if no file is specified,
into a directory that holds all users' crontabs.
The \-r option removes a user's crontab from the crontab directory.
\f2Crontab\fP \-l will list the crontab file for the
invoking user.
.PP
A user is permitted to use
.I crontab
if their name appears in the file
.B /usr/lib/cron/cron.allow.
If that file does not exist,
the file
.B /usr/lib/cron/cron.deny
is checked to determine if the user
should be denied access to
.I crontab.
If neither file exists, only root is allowed to
submit a job.
If either file is \f3at.deny\fR, global usage is permitted.
The allow/deny files consist of one user name
per line.
.PP
A crontab file
consists of lines of six fields each.
The fields are separated by spaces or tabs.
The first five are integer patterns that
specify the following:
.sp
.br
	minute (0\-59),
.br
	hour (0\-23),
.br
	day of the month (1\-31),
.br
	month of the year (1\-12),
.br
	day of the week (0\-6 with 0=Sunday).
.sp
Each of these patterns may
be either an asterisk  (meaning all legal values),
or a list of elements separated by commas.
An element is either a number, or
two numbers separated by a minus sign
(meaning an inclusive range).
Note that the specification of days
may be made by two fields
(day of the month and day of the week).
If both are specified as a list of elements, 
both are adhered to.
For example, 0 0 1,15 \(** 1 would run a command on the
first and fifteenth of each month, as well as on every Monday.
To specify days by only one field,
the other field should be set to \(**
(for example, 0 0 \(** \(** 1 would run a command only on Mondays).
.PP
The sixth field of a line in a crontab
file is a string that is executed by the shell 
at the specified times.
A percent character
in this field (unless escaped by 
\fB\e\fP)
is translated to a new-line
character.
Only the first line (up to a % or end of line)
of the command field is executed by the shell.
The other lines are made available to the
command as standard input.
.PP
The shell is invoked from your 
.B \s-1$HOME\s+1 
directory with an 
.BR arg0 " of " sh.
Users who desire to have their \fI.profile\fP executed must
explicitly do so in the crontab file.
.I Cron
supplies a default environment for every shell, defining
.BR \s-1HOME\s+1 ", " \s-1LOGNAME\s+1 ", " \s-1SHELL\s+1(=/bin/sh) ,
and 
.BR PATH(=:/bin:/usr/bin:/usr/lbin) .
.PP
.IR NOTE :
Users should remember to redirect the standard output 
and standard error of their commands!
If this is not done, any generated output or errors
will be mailed to the user.
.SH FILES
.TS
center;
l l.
/usr/lib/cron 	main cron directory
.br
/usr/spool/cron/crontabs 	spool area
.br
/usr/lib/cron/log 	accounting information
.br
/usr/lib/cron/cron.allow 	list of allowed users
.br
/usr/lib/cron/cron.deny 	list of denied users
.TE
.SH "SEE ALSO"
sh(1).
.br
cron(1M) in the \f2\s-1UNIX\s+1 System V Administrator Reference Manual\fR.
.\"	%W% of %G%


sysvr3/32/usr/src/cmd/cron/crontab.c
/*	Copyright (c) 1984 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#ident	"@(#)cron:crontab.c	1.8.1.4"
/**************************************************************************
 ***			C r o n t a b . c				***
 **************************************************************************

	date:	7/2/82
	description:	This program implements crontab (see cron(1)).
			This program should be set-uid to root.
	files:
		/usr/lib/cron drwxr-xr-x root sys
		/usr/lib/cron/cron.allow -rw-r--r-- root sys
		/usr/lib/cron/cron.deny -rw-r--r-- root sys

 **************************************************************************/


#include <sys/types.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <fcntl.h>
#include <ctype.h>
#include "cron.h"

#define TMPFILE		"_cron"		/* prefix for tmp file */
#define CRMODE		0444	/* mode for creating crontabs */

#define BADCREATE	"can't create your crontab file in the crontab directory."
#define BADOPEN		"can't open your crontab file."
#define BADSHELL	"because your login shell isn't /bin/sh, you can't use cron."
#define WARNSHELL	"warning: commands will be executed using /bin/sh\n"
#define BADUSAGE	"proper usage is: \n	crontab [file | -l| -r]"
#define INVALIDUSER	"you are not a valid user (no entry in /etc/passwd)."
#define NOTALLOWED	"you are not authorized to use cron.  Sorry."
#define EOLN		"unexpected end of line."
#define UNEXPECT	"unexpected character found in line."
#define OUTOFBOUND	"number out of bounds."

extern int opterr, optind, per_errno;
extern char *optarg, *xmalloc();
int err,cursor,catch();
char *cf,*tnam,line[CTLINESIZE];

main(argc,argv)
char **argv;
{
	int c, rflag, lflag, errflg;
	char login[UNAMESIZE],*getuser(),*strcat(),*strcpy();
	char *pp;
	FILE *fp;

	rflag = 0;
	lflag = 0;
	errflg = 0;
	while ((c=getopt(argc, argv, "lr")) != EOF)
		switch (c) {
			case 'l':
				argc--;
				if (rflag)
					errflg++;
				else
					lflag++;
				break;
			case 'r':
				argc--;
				if (lflag)
					errflg++;
				else
					rflag++;
				break;
			case '?':
				errflg++;
				break;
		}
	if (errflg || argc >2 || (argc == 2 && (rflag || lflag))) 
		crabort(BADUSAGE);
	pp = getuser(getuid());
	if(pp == NULL) {
		if (per_errno==2)
			crabort(BADSHELL);
		else
			crabort(INVALIDUSER); 
	}
	strcpy(login,pp);
	if (!allowed(login,CRONALLOW,CRONDENY)) crabort(NOTALLOWED);

	cf = xmalloc(strlen(CRONDIR)+strlen(login)+2);
	strcat(strcat(strcpy(cf,CRONDIR),"/"),login);
	if (rflag) {
		unlink(cf);
		sendmsg(DELETE,login);
		exit(0); 
	}
	if (lflag) {
		if((fp = fopen(cf,"r")) == NULL)
			crabort(BADOPEN);
		while(fgets(line,CTLINESIZE,fp) != NULL)
			fputs(line,stdout);
		fclose(fp);
		exit(0);
	}

	if (argc==1) copycron(stdin);
	else if (access(argv[1],04) || (fp=fopen(argv[1],"r"))==NULL) crabort(BADOPEN);
	else copycron(fp);
	sendmsg(ADD,login);
	if(per_errno == 2)
		fprintf(stderr,WARNSHELL);
	exit(0);
}


/******************/
copycron(fp)
/******************/
FILE *fp;
{
	FILE *tfp,*fdopen();
	char pid[6],*strcat(),*strcpy();
	int t;

	sprintf(pid,"%-5d",getpid());
	tnam=xmalloc(strlen(CRONDIR)+strlen(TMPFILE)+7);
	strcat(strcat(strcat(strcpy(tnam,CRONDIR),"/"),TMPFILE),pid);
	/* catch SIGINT, SIGHUP, SIGQUIT signals */
	if (signal(SIGINT,catch) == SIG_IGN) signal(SIGINT,SIG_IGN);
	if (signal(SIGHUP,catch) == SIG_IGN) signal(SIGHUP,SIG_IGN);
	if (signal(SIGQUIT,catch) == SIG_IGN) signal(SIGQUIT,SIG_IGN);
	if (signal(SIGTERM,catch) == SIG_IGN) signal(SIGTERM,SIG_IGN);
	if ((t=creat(tnam,CRMODE))==-1) crabort(BADCREATE);
	if ((tfp=fdopen(t,"w"))==NULL) {
		unlink(tnam);
		crabort(BADCREATE); 
	}
	err=0;	/* if errors found, err set to 1 */
	while (fgets(line,CTLINESIZE,fp) != NULL) {
		cursor=0;
		while(line[cursor] == ' ' || line[cursor] == '\t')
			cursor++;
		if(line[cursor] == '#')
			goto cont;
		if (next_field(0,59)) continue;
		if (next_field(0,23)) continue;
		if (next_field(1,31)) continue;
		if (next_field(1,12)) continue;
		if (next_field(0,06)) continue;
		if (line[++cursor] == '\0') {
			cerror(EOLN);
			continue; 
		}
cont:
		if (fputs(line,tfp) == EOF) {
			unlink(tnam);
			crabort(BADCREATE); 
		}
	}
	fclose(fp);
	fclose(tfp);
	if (!err) {
		/* make file tfp the new crontab */
		unlink(cf);
		if (link(tnam,cf)==-1) {
			unlink(tnam);
			crabort(BADCREATE); 
		} 
	}
	unlink(tnam);
}


/*****************/
next_field(lower,upper)
/*****************/
int lower,upper;
{
	int num,num2;

	while ((line[cursor]==' ') || (line[cursor]=='\t')) cursor++;
	if (line[cursor] == '\0') {
		cerror(EOLN);
		return(1); 
	}
	if (line[cursor] == '*') {
		cursor++;
		if ((line[cursor]!=' ') && (line[cursor]!='\t')) {
			cerror(UNEXPECT);
			return(1); 
		}
		return(0); 
	}
	while (TRUE) {
		if (!isdigit(line[cursor])) {
			cerror(UNEXPECT);
			return(1); 
		}
		num = 0;
		do { 
			num = num*10 + (line[cursor]-'0'); 
		}			while (isdigit(line[++cursor]));
		if ((num<lower) || (num>upper)) {
			cerror(OUTOFBOUND);
			return(1); 
		}
		if (line[cursor]=='-') {
			if (!isdigit(line[++cursor])) {
				cerror(UNEXPECT);
				return(1); 
			}
			num2 = 0;
			do { 
				num2 = num2*10 + (line[cursor]-'0'); 
			}				while (isdigit(line[++cursor]));
			if ((num2<lower) || (num2>upper)) {
				cerror(OUTOFBOUND);
				return(1); 
			}
		}
		if ((line[cursor]==' ') || (line[cursor]=='\t')) break;
		if (line[cursor]=='\0') {
			cerror(EOLN);
			return(1); 
		}
		if (line[cursor++]!=',') {
			cerror(UNEXPECT);
			return(1); 
		}
	}
	return(0);
}


/**********/
cerror(msg)
/**********/
char *msg;
{
	fprintf(stderr,"%scrontab: error on previous line; %s\n",line,msg);
	err=1;
}


/**********/
catch()
/**********/
{
	unlink(tnam);
	exit(1);
}


/**********/
crabort(msg)
/**********/
char *msg;
{
	fprintf(stderr,"crontab: %s\n",msg);
	exit(1);
}

/***********/
sendmsg(action,fname)
/****************/
char action;
char *fname;
{

	static	int	msgfd = -2;
	struct	message	*pmsg;

	pmsg = &msgbuf;
	if(msgfd == -2)
		if((msgfd = open(FIFO,O_WRONLY|O_NDELAY)) < 0) {
			if(errno == ENXIO || errno == ENOENT)
				fprintf(stderr,"cron may not be running - call your system administrator\n");
			else
				fprintf(stderr,"at: error in message queue open\n");
			return;
		}
	pmsg->etype = CRON;
	pmsg->action = action;
	strncpy(pmsg->fname,fname,FLEN);
	if(write(msgfd,pmsg,sizeof(struct message)) != sizeof(struct message))
		fprintf(stderr,"at: error in message send\n");
}
